docs/configure.md
# Config

When deploying the service, set configuration variables by including a `config.json` in the
root. Available configuration options:

- `timeout` _default `10000`_ - set the timeout used to render the target page.
- `port` _default `3000`_ - set the port to use for running and listening the rendertron service. Note if process.env.PORT is set, it will be used instead.
- `host` _default `0.0.0.0`_ - set the hostname to use for running and listening the rendertron service. Note if process.env.HOST is set, it will be used instead.
- `width` _default `1000`_ - set the width (resolution) to be used for rendering the page.
- `height` _default `1000`_ - set the height (resolution) to be used for rendering the page.
- `cache` _default `null`_ - set to `datastore` to enable caching on Google Cloud using datastore _only use if deploying to google cloud_, `memory` to enable in-memory caching or `filesystem` to enable disk based caching
- `cacheConfig` - an object array to specify caching options
- `renderOnly` - restrict the endpoint to only service requests for certain domains. Specified as an array of strings. eg. `['http://render.only.this.domain']`. This is a strict prefix match, so ensure you specify the exact protocols that will be used (eg. http, https).
- `closeBrowser`_default `false`_ - `true` forces the browser to close and reopen between each page render, some sites might need this to prevent URLs past the first one rendered returning null responses.
- `puppeteer` _default `{}`_ - an object to specify custom scripts to be executed during puppeteer page life cycle.

## cacheConfig

- `cacheDurationMinutes` _default `1440`_ - set an expiry time in minues, defaults to 24 hours. Set to -1 to disable cache Expiration
- `cacheMaxEntries` _default `100`_ - set the maximum number of entries stored in the selected cache method. Set to `-1` to allow unlimited caching. If using the datastore caching method, setting this value over `1000` may lead to degraded performance as the query to determine the size of the cache may be too slow. If you want to allow a larger cache in `datastore` consider setting this to `-1` and managing the the size of your datastore using a method like this [Deleting Entries in Bulk](https://cloud.google.com/datastore/docs/bulk-delete)
- `snapshotDir` _default `<your os's default tmp dir>/renderton`_ - **filesystem only** the directory the rendertron caches will be stored in

### Example

An example config file specifying a memory cache, with a 2 hour expiration, and a maximum of 50 entries

```javascript
{
    "cache": "memory",
    "cacheConfig": {
        "cacheDurationMinutes": 120,
        "cacheMaxEntries": 50
    }
}
```
## puppeteer config
Currently only supports:
  - `evaluateOnNewDocument` - inline script/function that is invoked after the document was created but before any of its scripts were run. You can use this as a hook to initialize/bootstrap any custom logic/state needed by the page. See [puppeteer.evaluateOnNewDocument](https://pptr.dev/#?product=Puppeteer&version=v8.0.0&show=api-pageevaluateonnewdocumentpagefunction-args)
  - `beforeSerialize` - path to a custom JS script that will be executed before the page is serialized by rendertron. The path is relative to rendertron root folder or can be absolute path.

### Example

Some CSS in JS frameworks (e.g styled-components), use insertRule to add styles to the page. These styles are not present in the DOM and don't make it into the cached HTML. Leading to issues like [styled-components#2577](https://github.com/styled-components/styled-components/issues/2577). You can use `evaluateOnNewDocument` to setup [global options](https://github.com/styled-components/styled-components/issues/2577#issuecomment-497815931) for these frameworks OR use `beforeSerialize` to inject [custom script](./test-resources/cssom-to-styles.js) to convert in-memory CSSOM to inline styles.

```javascript
"puppeteer": {
    "evaluateOnNewDocument": "SC_DISABLE_SPEEDY = true;",
    "beforeSerialize": "./cssom-to-styles.js"
}
```


src/config.ts
/*
 * Copyright 2018 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 not
 * use this file except in compliance with the License. You may obtain a copy
 of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 under
 * the License.
 */

'use strict';

import * as fse from 'fs-extra';
import * as path from 'path';
import * as os from 'os';

const CONFIG_PATH = path.resolve(__dirname, '../config.json');

export type Config = {
  cache: 'datastore' | 'memory' | 'filesystem' | null;
  cacheConfig: { [key: string]: string };
  timeout: number;
  port: string;
  host: string;
  width: number;
  height: number;
  reqHeaders: { [key: string]: string };
  headers: { [key: string]: string };
  puppeteerArgs: Array<string>;
  renderOnly: Array<string>;
  closeBrowser: boolean;
  clearCookies: boolean;
  restrictedUrlPattern: string | null;
};

export class ConfigManager {
  public static config: Config = {
    cache: null,
    cacheConfig: {
      snapshotDir: path.join(os.tmpdir(), 'rendertron'),
      cacheDurationMinutes: (60 * 24).toString(),
      cacheMaxEntries: '100',
    },
    timeout: 10000,
    port: '3000',
    host: '0.0.0.0',
    width: 1000,
    height: 1000,
    reqHeaders: {},
    headers: {},
    puppeteerArgs: ['--no-sandbox'],
    renderOnly: [],
    closeBrowser: false,
    restrictedUrlPattern: null,
    clearCookies: false,
  };

  static async getConfiguration(): Promise<Config> {
    // Load config.json if it exists.
    if (fse.pathExistsSync(CONFIG_PATH)) {
      const configJson = await fse.readJson(CONFIG_PATH);

      // merge cacheConfig
      const cacheConfig = Object.assign(
        ConfigManager.config.cacheConfig,
        configJson.cacheConfig
      );

      ConfigManager.config = Object.assign(ConfigManager.config, configJson);

      ConfigManager.config.cacheConfig = cacheConfig;
    }
    return ConfigManager.config;
  }
}
